### PR TITLE
ref(workflow): Optimize organization_group_index.py::get performance

### PR DESCRIPTION
**Pull request by iProgramStuff**
This is mostly a refactor for some performance optimizations. We recently added a lot of beefy queries to this to get additional stats (filtered and time range) for displaying in the list and now we're trying to improve it.

Previously, I believe this endpoint would make 3 serialize calls (getting lifetime, filtered, and time range), each making 1 call to Postgres, 2 snuba queries for seen stats (one for the count and one for times_seen, first_seen and last_seen), and 1 call to snuba for tsdb range query for stats. For a total of 3 postgres hits, 6 snuba queries, and 3 tsdb calls.

With this PR in it's current state, it now makes 1 serialize call, which makes 1 call to Postgres, 1 snuba query (the two queries for seen stats have been combined into one), and 2 tsdb calls (I don't call it for lifetime stats as its not needed). For a total of 1 postgres hit, 3 snuba queries, and 2 tsdb calls.

### CODE REVIEW

File: src/sentry/api/serializers/models/group.py [hunk 1]
```diff
@@ -713,7 +753,23 @@ def __init__(
...
 
+    def _get_seen_stats(self, item_list, user):
+        filtered_result = super(StreamGroupSerializerSnuba, self)._get_seen_stats(item_list, user)
+        # TODO: fix this hack. I would rather pass these vars down...but problems.
+        # At the very least, not losing their values after running this function would be nice.
+        self.snuba_filters = None
+        time_range_result = super(StreamGroupSerializerSnuba, self)._get_seen_stats(item_list, user)
+        self.start = None
+        self.end = None
+        lifetime_result = super(StreamGroupSerializerSnuba, self)._get_seen_stats(item_list, user)
```
Comment (markstory) [inline src/sentry/api/serializers/models/group.py:764]: Should this function take the start,end,snuba_fiilters values as parameters that have a default of None. If the value is None they fall back to the value stored in the attribute? That would let you get around this problem.

An alternative approach could be to move the work done in the parent class' `_get_seen_stats()` method into a helper method that doesn't use any instance state and then call that method here.

Comment (wedamija) [reply src/sentry/api/serializers/models/group.py:764]: +1 on extracting to a helper function

**Review by wedamija — COMMENTED** [review]
Could you also figure out roughly how many calls to snuba and tsdb (these are basically snuba as well) we were making before we the original changes? Just so that we can understand the extra load we'll add to snuba once we release.

I've been trying to figure out if there's a good way to make the `_get_seen_stats` query happen only once, but counting `unique_users` makes that hard.

I think at least some of these stats are only used on a hover over. One option (maybe a future optimization) could be to lazy load them on hover, or in a separate call after the page is loaded. The former would save us general load on snuba, and the latter would just save us initial page load time. Probably doesn't need to happen here - we can measure and see what the impact is.

File: src/sentry/api/serializers/models/group.py (hunk 1 above)
Comment (iProgramStuff) [reply src/sentry/api/serializers/models/group.py]: Brought it into a helper function. Going to try doing something similar for the get_stats -> query_tsdb chain.

**Comment by iProgramStuff** [comment]
> Could you also figure out roughly how many calls to snuba and tsdb (these are basically snuba as well) we were making before we the original changes? Just so that we can understand the extra load we'll add to snuba once we release.
> 
> I've been trying to figure out if there's a good way to make the `_get_seen_stats` query happen only once, but counting `unique_users` makes that hard.
> 
> I think at least some of these stats are only used on a hover over. One option (maybe a future optimization) could be to lazy load them on hover, or in a separate call after the page is loaded. The former would save us general load on snuba, and the latter would just save us initial page load time. Probably doesn't need to happen here - we can measure and see what the impact is.

I checked out eece42a1030e8b228f67cb53f4fc258f760adcda, which is a commit before Taylan merged in the first dynamic issue count work (on Tue Sep 1 14:34:59 - mostly noting for myself). It seems like it is calling serialize once, which makes two queries to get the seen stats (This PR has merged these into one), and then the 1 tsdb call. For a total of 1 postgres hit, 2 snuba queries, and 1 tsdb call. So getting the extra stats has added 1 snuba query, and 1 tsdb call but has made the snuba queries marginally more complicated by adding the unique count aggregation to them both.

I should be able to get some stats on page load times and performance from before + after as well.

File: src/sentry/api/serializers/models/group.py [hunk 2]
```diff
@@ -467,6 +467,15 @@ def serialize(self, obj, attrs, user):
             "hasSeen": attrs["has_seen"],
             "annotations": attrs["annotations"],
         }
+        +self._convert_seen_stats(attrs)
```
Comment (wedamija) [inline src/sentry/api/serializers/models/group.py:470]: This formatting is really weird - did black run here?

File: src/sentry/api/serializers/models/group.py [hunk 3]
```diff
@@ -638,42 +644,74 @@ def serialize(self, obj, attrs, user):
...
+        "query",
+        "status",
+        "bookmarked_by",
+        "assigned_to",
+        "unassigned",
+        "subscribed_by",
+        "active_at",
+        "first_release",
+        "first_seen",
+        "message",
```
Comment (wedamija) [inline src/sentry/api/serializers/models/group.py:657]: Probably worth removing `message` based on our discussion this morning

File: src/sentry/api/serializers/models/group.py [hunk 4]
```diff
@@ -713,21 +762,41 @@ def __init__(
...
+        partial_execute_seen_stats_query = functools.partial(
+            self._execute_seen_stats_query,
+            item_list=item_list,
+            environment_ids=self.environment_ids,
+            start=self.start,
+            end=self.end,
+        )
+        filtered_result = partial_execute_seen_stats_query(conditions=self.conditions)
+        time_range_result = partial_execute_seen_stats_query()
+        lifetime_result = partial_execute_seen_stats_query(start=None, end=None)
```
Comment (wedamija) [inline src/sentry/api/serializers/models/group.py:775]: I actually didn't realise you could override the values of already passed args in a partial like this and had to verify, looks good!

**Review by wedamija — APPROVED** [review]
> > Could you also figure out roughly how many calls to snuba and tsdb (these are basically snuba as well) we were making before we the original changes? Just so that we can understand the extra load we'll add to snuba once we release.
> > I've been trying to figure out if there's a good way to make the `_get_seen_stats` query happen only once, but counting `unique_users` makes that hard.
> > I think at least some of these stats are only used on a hover over. One option (maybe a future optimization) could be to lazy load them on hover, or in a separate call after the page is loaded. The former would save us general load on snuba, and the latter would just save us initial page load time. Probably doesn't need to happen here - we can measure and see what the impact is.
> 
> I checked out [eece42a](https://github.com/getsentry/sentry/commit/eece42a1030e8b228f67cb53f4fc258f760adcda), which is a commit before Taylan merged in the first dynamic issue count work (on Tue Sep 1 14:34:59 - mostly noting for myself). It seems like it is calling serialize once, which makes two queries to get the seen stats (This PR has merged these into one), and then the 1 tsdb call. For a total of 1 postgres hit, 2 snuba queries, and 1 tsdb call. So getting the extra stats has added 1 snuba query, and 1 tsdb call and has also made the snuba queries marginally more complicated by adding the unique count aggregation to them both.
> 
> I should be able to get some stats on page load times and performance from before + after as well.

Ok, that's good info. I think that will be fine, and we can optimize further if we get either page slowness, or complaints from search and storage about additional load on Clickhouse.

This is looking really good, I appreciate all the work you've done to refactor this!

File: src/sentry/api/serializers/models/group.py (hunk 3 above)
Comment (iProgramStuff) [reply src/sentry/api/serializers/models/group.py:657]: Removed :-)

File: src/sentry/api/serializers/models/group.py (hunk 4 above)
Comment (iProgramStuff) [reply src/sentry/api/serializers/models/group.py:775]: Haha yeah, me too. Liked this better than passing start + end to filtered + time_range :-)

File: src/sentry/api/serializers/models/group.py (hunk 2 above)
Comment (iProgramStuff) [reply src/sentry/api/serializers/models/group.py]: I moved this addition to the other side of the dict...does that feel less awkward?

File: src/sentry/api/serializers/models/group.py [hunk 5]
```diff
@@ -722,22 +771,56 @@ def __init__(
...
+                else None
+            )
+            lifetime_result = (
+                partial_execute_seen_stats_query(start=None, end=None)
+                if self.start or self.end
+                else None
+            )
+            for item in item_list:
+                time_range_result[item].update({"filtered": filtered_result.get(item)})
+                time_range_result[item].update({"lifetime": lifetime_result.get(item)})
```
Comment (markstory) [inline src/sentry/api/serializers/models/group.py:798]:
```suggestion
                time_range_result[item].update({
                    "filtered": filtered_result.get(item),
                    "lifetime": lifetime_result.get(item),
                })
```
You could combine these updates.

File: src/sentry/api/serializers/models/group.py [hunk 6]
```diff
@@ -751,4 +834,11 @@ def serialize(self, obj, attrs, user):
         if self.matching_event_id:
             result["matchingEventId"] = self.matching_event_id
 
+        if self.has_dynamic_issue_counts:
+            if self.stats_period:
+                attrs["lifetime"].update({"stats": None})  # Not needed in current implentation
```
Comment (markstory) [inline src/sentry/api/serializers/models/group.py:839]:
```suggestion
                attrs["lifetime"].update({"stats": None})  # Not needed in current implementation
```

**Pull request merged**